### ISSUE TITLE
For `in` parameters non-virtual methods/delegates/etc, they should have modopt(In)

### ISSUE THREAD

**Issue by OmarTawfik** [dotnet/roslyn#22642]
If you're a library author, and you already have such signature:

```
class API
{
   void Method(ref int x) { }
}
```

And you want to make use of the new `in` parameters, you cannot introduce an overload in the new API version that let's your users use it. Because they would be identical in IL.
Compiler should put the `modopt(In)` on signatures it doesn't put a `modreq(In)` on to enable overloading.

cc @tannergooding @jcouv

**Comment by tannergooding** [comment]
The downside to not doing this is that converting an existing signature from `ref` to `in` is a source breaking change.

**Comment by jcouv** [comment]
To capture Tanner's scenarios:
1. You have a shipped `ref` API and then you add an `in` overload. Without OP, we can never enable this scenario. With OP fixed, we could update the language to allow such overloads.
2. You have a shipped `ref` API and you replace it with an `in` API. Without OP, this is a source break to consumers of the API. With OP, this is a binary break.

So the benefit of OP is that is would allow such overloading in the future and that would be the recommended path for upgrading `ref` APIs that effectively treat those refs as readonly.

Tagging @VSadov

**Comment by jcouv** [comment]
From discussion with @jaredpar, we won't do this. A couple of reasons:
- we had similar API issues for people switching from `ref` to `out`, which we accepted.
- if we really really want to do this, he thought we could add the modopt later (but we got interrupted in discussion).
- he raised a point about complexity of determining when a method is virtual (there are some optimization cases that introduce subtleties). But Omar tested that the optimization isn't done anymore and anyways, the complexity affects the modreq case, not the modopt case. The modopt case would just be whenever we were not going to put that modreq.
